src/libraries/Microsoft.Extensions.Http/src/DefaultHttpMessageHandlerBuilder.cs

```diff
@@ -44,5 +52,36 @@ public override HttpMessageHandler Build()
 
             return CreateHandlerPipeline(PrimaryHandler, AdditionalHandlers);
         }
+
+#pragma warning disable CA1822, CA1859 // Mark members as static, Use concrete types when possible for improved performance
+        private HttpMessageHandler CreatePrimaryHandler()
+#pragma warning restore CA1822, CA1859
+        {
+#if NET
+            // On platforms where SocketsHttpHandler is supported, HttpClientHandler is a thin wrapper
+            // around it. By using SocketsHttpHandler directly, we can avoid the overhead of the wrapper,
+            // but more importantly, we can configure it to limit the lifetime of its pooled connections
+            // to match the requested lifetime of the handler itself. That way, if/when someone holds on
+            // to a resulting HttpClient for a prolonged period of time, it'll still benefit from connection
+            // recycling, and without needing to tear down and reconstitute the rest of the handler pipeline.
+            if (SocketsHttpHandler.IsSupported)
+            {
+                SocketsHttpHandler handler = new();
+
+                if (Services.GetService<IOptionsMonitor<HttpClientFactoryOptions>>() is IOptionsMonitor<HttpClientFactoryOptions> optionsMonitor)
```

Comment:
I need to think more about this part. I need to check whether it would give us an expected result in all circumstances. Also I need to make sure the `_name` is also always set to an expected value.

I'll return to it tomorrow.

Reply:
Sorry I still don't have an answer to that. It's in my pipeline, I will reply by Monday.

Reply:
Ok, so the answers:
- `_name` is set in the default factory right after creating the instance and before any of the actions are applied to the builder. So we're ok from this regard.
- getting the options from IOptionsMonitor here in this point in time (while creating the handler pipeline) also aligns with the options usage in other places, including the default factory, so it's ok as well.

👍

Reply:
@stephentoub Is the check `is IOptionsMonitor<HttpClientFactoryOptions> optionsMonitor` only needed to verify whether the options could be resolved from the container in general? If so, the check is redundant, since the instance of `DefaultHttpClientFactory` was already created by this point, so `IOptionsMonitor<HttpClientFactoryOptions>` was already successfully injected into its constructor.

I'd vote for simply injecting `IOptionsMonitor<HttpClientFactoryOptions>` into `DefaultHttpMessageHandlerBuilder`'s constructor here as well.

Reply:
> Is the check is IOptionsMonitor<HttpClientFactoryOptions> optionsMonitor only needed to verify whether the options could be resolved from the container in general?

Mainly*, yes. It's guaranteed to have been registered? Where does that happen?

*_Since IServiceProvider was being injected into the constructor rather than individual resources, this also seemed to be the more consistent approach, but I can change it if desired._

Reply:
> Since IServiceProvider was being injected into the constructor

It's there because we need to flow it to configuration callbacks, bc we don't really know what services the callbacks will need at this point.

> It's guaranteed to have been registered? Where does that happen?

Yes, it happens in the "base" `AddHttpClient()` method which is called from all other `AddHttpClient` methods:
https://github.com/dotnet/runtime/blob/c6bfb06e23458e583c46cb2162c827acc8dd2f35/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs#L31
which in turn has
https://github.com/dotnet/runtime/blob/c6bfb06e23458e583c46cb2162c827acc8dd2f35/src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs#L28

Constructor injection will align with the other usages in
https://github.com/dotnet/runtime/blob/c6bfb06e23458e583c46cb2162c827acc8dd2f35/src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs#L65
and https://github.com/dotnet/runtime/blob/c6bfb06e23458e583c46cb2162c827acc8dd2f35/src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandlerBuilderFilter.cs#L23

So I believe it will be better to do it the same way, that being said -- I don't want to hold back this PR only for that change; I can do it myself later as a follow-up.